Allow getting all backend names

cccclai opened the pull request:
Summary: Allow getting all backends name in both python and c++

Differential Revision: D69691354

larryliu0820 commented:
We are moving to `[runtime/__init__.py](https://github.com/pytorch/executorch/blob/main/runtime/__init__.py#L136)` for our runtime python APIs. Can you add this new API there as well?

larryliu0820 reviewed (changes requested):
Put it back to your queue

dbort commented on runtime/backend/interface.cpp:
```diff
@@ -55,5 +55,9 @@ Error register_backend(const Backend& backend) {
   return Error::Ok;
 }
 
+Span<const Backend> get_registered_backends() {
```
If we ever allow for runtime-registered (or deregistered) backends, this API will be very hard to support: the function must return a single, contiguous buffer containing all backends, and the caller could maintain a pointer to that buffer for the lifetime of the process.

It would be safer and more maintainable to have an API like
```
size_t get_num_registered_backends();
Result<const char*> get_backend_name(size_t index);
```

Also I think we should hide/remove the current `Backend` struct, which is really an internal detail of how the table is stored. It's used to register backends, but that actually complicates things for callers since they need to create that struct. It'd be easier for callers if the API were `register_backend(const char* name, BackendInterface* backend);`

cccclai commented:
@larryliu0820 @dbort updated and addressed the comments. Please take another look.

larryliu0820 reviewed (approved):
Thanks for addressing comments

GregoryComer commented:
This could be done as a follow-up as I wouldn't want to block this diff, but it would be nice to be able to get only available backends - i.e. backends where is_available() returns true. This is useful for determining if the actual hardware is present. For example, the Qualcomm backend might be linked but not running on a device that actually has a supported NPU. The older API that returned the Backend directly would have allowed for calling is_available(), but the current approach does not. That's fine, as we can add another API to do this - just a consideration.

GregoryComer commented on extension/pybindings/pybindings.cpp:
```diff
@@ -975,6 +979,19 @@ py::list get_operator_names() {
+    auto backend_name_res = get_backend_name(i);
+    THROW_IF_ERROR(backend_name_res.error(), "Failed to get backend name");
+    auto backend_name = backend_name_res.get();
+    printf("backend name: %s\n", backend_name);
```
Maybe leftover debug print?

cccclai commented:
> This could be done as a follow-up as I wouldn't want to block this diff, but it would be nice to be able to get only available backends - i.e. backends where is_available() returns true. This is useful for determining if the actual hardware is present. For example, the Qualcomm backend might be linked but not running on a device that actually has a supported NPU. The older API that returned the Backend directly would have allowed for calling is_available(), but the current approach does not. That's fine, as we can add another API to do this - just a consideration.

oh I actually already have this diff as part of the follow up. Just haven't submitted yet.

mergennachin commented on runtime/core/error.h:
```diff
@@ -93,6 +93,9 @@ enum class Error : error_code_t {
 
+  /// Backend registration: trying to access an invalid index in the backend
+  /// registration table.
+  BackendTableAccessFail = 0x33,
```
I wouldn't add a new error message, just for this. InvalidArgument = 0x12 is sufficient for this case.

cccclai replied:
Sure

mergennachin commented on runtime/backend/interface.cpp:
```diff
@@ -55,5 +55,16 @@ Error register_backend(const Backend& backend) {
+
+Result<const char*> get_backend_name(size_t index) {
+  if (index >= num_registered_backends) {
+    return Error::BackendTableAccessFail;
```
InvalidArgument = 0x12

mergennachin commented on runtime/__init__.py:
```diff
@@ -125,6 +125,21 @@ def load_method(self, name: str) -> Optional[Method]:
+        self._legacy_module = legacy_module
+
+    @property
+    def _get_backend_names(self) -> Set[str]:
```
why start with _?

cccclai replied:
updated

mergennachin commented on runtime/__init__.py:
```diff
@@ -125,6 +125,21 @@ def load_method(self, name: str) -> Optional[Method]:
+    @property
+    def _get_backend_names(self) -> Set[str]:
+        """
+        Returns the names of all registered backends as a list of strings.
```
as a set?

cccclai replied:
list is better. Borrowed some code from `get_operators_names` but now I don't think we will have duplication

mergennachin commented on extension/pybindings/test/test_backend_pybinding.py:
```diff
@@ -0,0 +1,12 @@
+    def test_backend_name_list(
+        self,
+    ) -> None:
+        all_backend_name = _get_backend_names()
```
In the unit test, use the new pybindings API

```
from executorch.runtime import Runtime

runtime = Runtime.get()

runtime.backend_registry.get_backend_names()
```

cccclai replied:
updated

mergennachin commented on runtime/__init__.py:
```diff
@@ -125,6 +125,21 @@ def load_method(self, name: str) -> Optional[Method]:
+        self._legacy_module = legacy_module
+
+    @property
+    def _get_backend_names(self) -> Set[str]:
```
I'd rename to `get_registered_backends()`

mergennachin commented on extension/pybindings/pybindings.cpp:
```diff
@@ -975,6 +979,18 @@ py::list get_operator_names() {
   return res;
 }
 
+py::list get_backend_names() {
```
get_registered_backends

cccclai replied:
updated to `get_registered_backend_names`

mergennachin commented on extension/pybindings/pybindings.cpp:
```diff
@@ -975,6 +979,18 @@ py::list get_operator_names() {
 
+py::list get_backend_names() {
+  size_t n_of_registered_backends = get_num_registered_backends();
+  py::list res;
```
is the result list or set? in the typehint, it is Set[str] but you're returning a list. Which one is it?

cccclai replied:
It should be list

mergennachin commented on extension/pybindings/pybindings.cpp:
```diff
@@ -87,10 +88,13 @@ using ::executorch::extension::BufferDataLoader;
 using ::executorch::extension::MallocMemoryAllocator;
 using ::executorch::extension::MmapDataLoader;
 using ::executorch::runtime::ArrayRef;
+using ::executorch::runtime::Backend;
```
do you need this?

cccclai replied:
updated

dbort commented on extension/pybindings/pybindings.cpp:
```diff
@@ -1028,6 +1043,10 @@ PYBIND11_MODULE(EXECUTORCH_PYTHON_MODULE_NAME, m) {
       },
       call_guard);
+  m.def(
+      "_get_registered_backend_names",
```
Please add this new function to https://github.com/pytorch/executorch/blob/main/extension/pybindings/pybindings.pyi

cccclai replied:
updated

dbort reviewed (approved):
Thanks for changing the API to be index-based. Looks good apart from my final request to update the .pyi file.

cccclai committed:
Allow getting all backend names (#8520)

Summary:

Allow getting all backends name in both python and c++

Reviewed By: omerjerk

Differential Revision: D69691354
